### DISCUSSION TITLE
EdgeQL queries

### DISCUSSION THREAD

**Discussion by nsidnev** [geldata/gel#1665] [Show and tell]
Hi!

I made a small library that can help deal with complex EdgeQL queries without using raw python strings. You put your queries somewhere in the repository and then "init" them from python code. Once initialized, you can call these queries just like any other python code.

This approach is quite popular, for example, there are the following similar libraries for python and other languages:
* Async and sync python >= 3.6 - [nackjicholson/aiosql](https://github.com/nackjicholson/aiosql)
* Sync python < 3.6 - [honza/anosql](https://github.com/honza/anosql)
* Closure - [krisajenkins/yesql](https://github.com/krisajenkins/yesql)
* Erlang - [artemeff/eql](https://github.com/artemeff/eql)

The main reasons for this approach can be found here: https://nackjicholson.github.io/aiosql/getting-started/ .
But in short:
* Writing compicated queries even with the query builder will never be easy.
* Raw strings in python code can't be easily integrated with LSP and other tools(even PyCharm Professional, which has tight database integration and can manipulate raw strings like SQL queries, has some problems).
* This approach can be easily combined with query builders to achieve improved development.

Link to the repository: https://github.com/nsidnev/edgeql-queries
Documentation: https://nsidnev.github.io/edgeql-queries/

**Comment by 1st1** [comment]
Awesome, this looks great!

**Reply by 1st1** [reply]
I'd also look into using YAML, or pure Python (with metaclasses), or even "one query per file approach" to avoid the need to lex edgeql.

**Reply by nsidnev** [reply]
> avoid the need to lex edgeql

Do you mean usage of `edgeql_lexer` package? It is only used to build documentation as `pygments` does not currently support EdgeQL out of the box :(

Queries are parsed with regex by `# name: <query_name>` comment line and then grouped together by directory names (or in the root instance of `Queries`, if no subdirectories exist).

This way "one query per file approach" still can be used. I used it together with `aiosql`, placing simple queries in a common file(`users.sql` for example), and splitting complex queries into several(`chech_user_can_subscribe_on_author.sql` or other). And then I combined all `.sql` files in directories by their domain(`sql/users/`).

**Reply by 1st1** [reply]
> Do you mean usage of edgeql_lexer package? It is only used to build documentation as pygments does not currently support EdgeQL out of the box :(

Ah, got it!

> Queries are parsed with regex by # name: <query_name> comment line and then grouped together by directory names (or in the root instance of Queries, if no subdirectories exist).

Makes sense.

I really like the idea, btw. We'll see if we can use this in one of our internal projects and give you more feedback.
